dc/s2n-quic-dc/events/connection.rs

```diff
@@ -1,26 +1,209 @@
+
+#[event("stream:write_socket_blocked")]
+pub struct StreamWriteSocketBlocked {
+    /// The number of bytes that the stream tried to write to the socket
+    #[measure("provided", Bytes)]
+    provided_len: usize,
+
+    /// The amount of time it took to process the write request
+    ///
+    /// Note that this includes both any syscall and encryption overhead
+    #[measure("processing_duration", Duration)]
+    processing_duration: core::time::Duration,
+}
+
+#[event("stream:write_socket_errored")]
+pub struct StreamWriteSocketErrored {
+    /// The number of bytes that the stream tried to write to the socket
+    #[measure("provided", Bytes)]
+    provided_len: usize,
+
+    /// The amount of time it took to process the write request
+    #[measure("processing_duration", Duration)]
+    processing_duration: core::time::Duration,
+
+    /// The system `errno` from the returned error
+    errno: Option<i32>,
 }
 
-#[event("application:read")]
-pub struct ApplicationRead {
+#[event("stream:read_flushed")]
+pub struct StreamReadFlushed {
     /// The number of bytes that the application tried to read
     #[measure("capacity", Bytes)]
     capacity: usize,
 
-    /// The amount that was read
+    /// The amount that was read into the provided buffer
     #[measure("committed", Bytes)]
     #[counter("committed.total", Bytes)]
-    read_len: usize,
+    committed_len: usize,
+
+    /// The amount of time it took to process the read request
+    ///
+    /// Note that this includes both any syscall and decryption overhead
+    #[measure("processing_duration", Duration)]
+    processing_duration: core::time::Duration,
+}
+
+#[event("stream:read_blocked")]
+pub struct StreamReadBlocked {
+    /// The number of bytes that the application tried to read
+    #[measure("capacity", Bytes)]
+    capacity: usize,
+
+    /// The amount of time it took to process the read request
+    ///
+    /// Note that this includes both any syscall and decryption overhead
+    #[measure("processing_duration", Duration)]
+    processing_duration: core::time::Duration,
+}
+
+#[event("stream:read_errored")]
+pub struct StreamReadErrored {
+    /// The number of bytes that the application tried to read
+    #[measure("capacity", Bytes)]
+    capacity: usize,
+
+    /// The amount of time it took to process the read request
+    ///
+    /// Note that this includes both any syscall and decryption overhead
+    #[measure("processing_duration", Duration)]
+    processing_duration: core::time::Duration,
+
+    /// The system `errno` from the returned error
+    errno: Option<i32>,
+}
+
+#[event("stream:read_shutdown")]
+pub struct StreamReadShutdown {
+    /// If the stream required a background task to drive the stream shutdown
+    #[bool_counter("background")]
+    background: bool,
+}
+
+#[event("stream:read_socket_flushed")]
+pub struct StreamReadSocketFlushed {
+    /// The number of bytes that the stream tried to read from the socket
+    #[measure("capacity", Bytes)]
```

Comment:
This buffer is always an internal buffer IIRC, right? Do we expect meaningful size differences on that?

Reply:
Yeah it's always an internal buffer. I was curious to see if we were ever needing to read more for stream sockets to complete a packet, in which case this buffer would be undersized.